### ISSUE TITLE
Doubly freed pointer in torch::cat error handling when called via pybind11.

### ISSUE THREAD

**Issue by heiner** [pytorch/pytorch#18283]
## 🐛 Bug

Calling `torch::cat` with an illegal concatenation dimension in a pybind11-exposed C++ function results in a malloc error:

```
python3(47071,0x1147385c0) malloc: *** error for object 0x7f9e916be8d8: pointer being freed was not allocated
python3(47071,0x1147385c0) malloc: *** set a breakpoint in malloc_error_break to debug
```

## To Reproduce

1. Clone the code in the gist at https://gist.github.com/heiner/c03aa5df92408974922aea7ec499405c via `git clone https://gist.github.com/c03aa5df92408974922aea7ec499405c.git catbug`.
2. Run `cd catbug; python3 setup.py build develop`.
3. Run `python3 run.py`.

The relevant point in the backtrace is at
```
frame #4: 0x0000000119063f94 torchloops.cpython-37m-darwin.so`pybind11::cpp_function::dispatcher(self=0x0000000118df7b10, args_in=0x0000000118df5f60, kwargs_in=0x0000000000000000) at pybind11.h:691
   688 	            }
   689 	            PyErr_SetString(PyExc_SystemError, "Exception escaped from default exception translator!");
   690 	            return nullptr;
-> 691 	        }
   692
   693 	        auto append_note_if_missing_header_is_suspected = [](std::string &msg) {
   694 	            if (msg.find("std::") != std::string::npos) {
```

## Expected behavior

The underlying `c10::Error` should be raised in C++ and surfaced to the user.

## Environment

```
$ python3 collect_env.py
Collecting environment information...
PyTorch version: 1.0.0.dev20190313
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14.3
GCC version: Could not collect
CMake version: version 3.12.2

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.2
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.1                      144
[conda] mkl-include               2019.1                      144
[conda] mkl_fft                   1.0.10           py37h5e564d8_0
[conda] mkl_random                1.0.2            py37h27c97d8_0
[conda] pytorch-nightly           1.0.0.dev20190313         py3.7_0    pytorch
[conda] torchgym                  0.0.0                     dev_0    <develop>
[conda] torchgymatari             0.0.0                    pypi_0    pypi
```

**Comment by soumith** [comment]
cc: @yf225 @gchanan

**Comment by skrah** [comment]
@heiner On Linux using the latest master (9005a2c0fc) I cannot reproduce this.  I'm getting a `c10::Error` and the test case also passes `Valgrind` --- apart from some `definitely lost` leaks, some of them in external packages, but those are a different issue.

**Comment by skrah** [comment]
The expected error:

```
RuntimeError: Dimension out of range (expected to be in range of [-1, 0], but got 10) (maybe_wrap_dim at /home/stefan/pytorch/c10/core/WrapDimMinimal.h:20)
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x6a (0x7f9fbc8fd5ca in /home/stefan/rel/lib/python3.8/site-packages/torch/lib/libc10.so)
frame #1: <unknown function> + 0x7ded39 (0x7f9fbd4f6d39 in /home/stefan/rel/lib/python3.8/site-packages/torch/lib/libcaffe2.so)
```

**Comment by gchanan** [comment]
perhaps this was an ABI issue?  I'm going to close because that's my best guess and because of the lack of reproduction reported by @skrah.  @heiner can you kindly reopen if you disagree?

**Comment by heiner** [comment]
I don't believe this is an ABI issue as it occurs on Mac using clang (although incidentally cpp_extension.py prints its warning message unless I set CXX=c++, which I believe doesn't actually change which compiler gets used):

```$ c++ --version
Apple LLVM version 10.0.0 (clang-1000.11.45.5)
Target: x86_64-apple-darwin18.5.0
Thread model: posix
InstalledDir: /Applications/Xcode_10.1_fb.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

The issue still occurs for me, on MacOS with the latest nightly conda package (`1.1.0.dev20190507-py3.7_0`).

**Comment by skrah** [comment]
With `conda` there are sometimes stale libraries in e.g. `miniconda3/lib/*` that cause issues like this.

I don't have OS X, so I quickly setup a test on Travis CI with `pytorch-nightly` that passes (i.e. produces a  `c10::Error`) with a pristine `conda` environment on Mac OS X:

https://travis-ci.org/skrah/test-osx-issue/builds/529914165#L720

**Comment by skrah** [comment]
The script is here:

https://github.com/skrah/test-osx-issue/blob/master/.travis.yml

**Comment by heiner** [comment]
Thanks for investigating Stefan! I'm still not sure what causes this, but given that it works in a fresh environment I am happy to have this closed.

**Comment by skrah** [comment]
Yes, it's hard to tell.  With another package I had issues similar to this one; the causes were either stale libraries in system directories or in `miniconda/lib`.

I'm closing then, of course please reopen if a fresh `miniconda` install does not work either.

**Comment by MarcoSelvatici** [comment]
Hi,
I encountered the same issue today (https://github.com/fairinternal/hackrl/pull/5), and I could reproduce it on a fresh conda environment following the steps proposed in the [description](https://github.com/pytorch/pytorch/issues/18283#issue-423870438).
I am using MacOs:
```
ProductName:	Mac OS X
ProductVersion:	10.14.5
```

**Comment by heiner** [comment]
As @MarcoSelvatici says we were bitten by this issue again today. This issue can make debugging other bugs tricky as `c10::Error` doesn't inherit from `std::exception`, so one has to guess what's wrong in order to catch the right exception. In our most recent case it was that tensors of different dtypes were fed into our batching system.

I'm not sure how to try to reproduce this except that it seems to happen on a standard mac with anaconda. I suppose we could look at @skrah's Travis CI setup above.

**Comment by heiner** [comment]
> perhaps this was an ABI issue? I'm going to close because that's my best guess and because of the lack of reproduction reported by @skrah. @heiner can you kindly reopen if you disagree?

@gchanan I'm not sure I agree or disagree but at any rate I don't think I can reopen this issue at this point?

**Comment by gchanan** [comment]
@yf225 can you see if you can reproduce this?

**Comment by heiner** [comment]
Hey @yf225: I start to suspect this is an issue relating to some kind of binary incompatibility. Have you tried reproducing the steps above on a MacBook after installing Pytorch via `pip install torch`?

**Comment by heiner** [comment]
> This issue can make debugging other bugs tricky as `c10::Error` doesn't inherit from `std::exception`, so one has to guess what's wrong in order to catch the right exception.

Notice that I was wrong here but nobody paid enough attention to correct this. `c10::Error` is in fact a `public std::exception`. But when using PyTorch via the pre-built wheel on MacOS one cannot catch these errors with `catch (const std::exception &e)`. However non-portable code like this

```
try {
  // blow up stuff e.g. via a wrong dtype
} catch (...) {
  std::cout << "... exception" << std::endl;
  int status;
  std::cout << abi::__cxa_demangle(
                    abi::__cxa_current_exception_type()->name(), 0, 0,
                    &status)
            << std::endl;
}
```

will give a hint to what the actual problem was. 

This is another incidence of a weird bug showing up when installing PyTorch the standard way and building C++ PyTorch extensions. I'd love for someone to finally fix this.

**Comment by ezyang** [comment]
Usually when this happens, it means we've got two copies of libstdc++ in your address space, and so the two "exceptions" aren't comparing equal per RTTI.

**Comment by ezyang** [comment]
There's more fine grained control over RTLD_GLOBAL in PyTorch 1.5, which should have made this better. I'm going to optimistically downgrade the priority here, subject to user activity
